plugins/bcm57xx/fu-bcm57xx-device.c

```diff
@@ -52,6 +51,66 @@ fu_bcm57xx_device_probe(FuDevice *device, GError **error)
 	return fu_udev_device_set_physical_id(FU_UDEV_DEVICE(device), "pci", error);
 }
 
+#ifdef HAVE_ETHTOOL_H
+static gboolean
+fu_bcm57xx_device_submit_ifreq(FuBcm57xxDevice *self, guint8 *buf, gsize bufsz, GError **error)
+{
+	FuDeviceEvent *event = NULL;
+	struct ifreq ifr = {0};
+	g_autofree gchar *event_id = NULL;
+
+	g_return_val_if_fail(buf != NULL, FALSE);
+
+	/* emulated */
+	if (fu_device_has_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_EMULATED) ||
+	    fu_context_has_flag(fu_device_get_context(FU_DEVICE(self)),
+				FU_CONTEXT_FLAG_SAVE_EVENTS)) {
+		g_autofree gchar *buf_base64 = g_base64_encode(buf, bufsz);
+		event_id = g_strdup_printf("SiocethtoolIoctl:"
+					   "Data=%s,"
+					   "Length=0x%x",
+					   buf_base64,
+					   (guint)bufsz);
+	}
+
+	/* emulated */
+	if (fu_device_has_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_EMULATED)) {
+		event = fu_device_load_event(FU_DEVICE(self), event_id, error);
+		if (event == NULL)
+			return FALSE;
+		return fu_device_event_copy_data(event, "DataOut", buf, bufsz, NULL, error);
+	}
+
+	/* save */
+	if (fu_context_has_flag(fu_device_get_context(FU_DEVICE(self)),
+				FU_CONTEXT_FLAG_SAVE_EVENTS)) {
+		event = fu_device_save_event(FU_DEVICE(self), event_id);
+		fu_device_event_set_data(event, "Data", buf, bufsz);
+	}
+
+	/* we can't use the emulation support in fu_udev_device_ioctl() as
+	 * the buffer is specified indirectly using ifr_data */
```

Comment:
I _think_ this is the same problem with why amdgpu probing can't work either in emulation currently.

Reply:
Ya, the return_pointer stuff will need to be handled in the same way -- but there are other places it's going to fail too; e.g. using g_file_tests on random sysfs files -- we should add a fu_udev_device helper for that so we can emulate it. That'd work for bcm57xx too. The other pain point is amdgpu_device_initialize() but we could wrap that too. I've got an early AMD thinkpad here; goes that have anything new enough for pspvbflash or perhaps you could send me something PVTish that I could use to help?

Reply:
1/3: https://github.com/fwupd/fwupd/pull/7972